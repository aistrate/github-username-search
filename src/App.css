.App {
  max-width: 50rem;
  min-width: 20rem;
  min-height: 100vh;
  margin: auto;
}

.App__header {
  background-color: #1b4371;
  color: white;
  font-size: 2.25rem;
  text-align: center;
  padding: 0.6em 0;
}

@media (max-width: 50em) {
  .App__header {
    font-size: calc(1.25rem + 2vw);
  }
}

.Nav {
  list-style-type: none;
  margin: 0;
  padding: 0;
  overflow: hidden;
  background-color: rgb(0, 0, 0, 0.85);
  font-size: 1.25rem;
}

.NavItem {
  float: left;
}

.NavItem__link {
  display: block;
  color: rgb(255, 255, 255, 0.5);
  font-weight: 600;
  padding: 0.4em 0.8em;
  text-align: center;
  text-decoration: none;
}

.NavItem__link:hover {
  color: white;
  background-color: rgb(255, 255, 255, 0.1);
}

.NavItem__link--active {
  color: white;
}

.Content {
  padding: 1rem;
  overflow-x: hidden;
}

.SearchForm {
  margin-top: 1.5rem;
}

.SearchForm__instructions {
  margin-top: 0.6rem;
  opacity: 0.8;
}

.TextInput {
  font-family: inherit;
  font-size: 1rem;
  font-weight: 600;
  color: #0e1e25;
  border: 1px solid #a2a9b1;
  border-radius: 6px;
  outline: none;
  width: 12em;
  height: 1.5em;
  padding: 7px 14px;
}

.TextInput::placeholder {
  color: darkgray;
}

.TextInput:hover {
  border-color: #72777d;
}

.TextInput:focus {
  border-color: #3366cc;
  box-shadow: inset 0 0 0 1px #3366cc;
}

@media (max-width: 40em) {
  .TextInput {
    width: 10em;
  }
}

.Button {
  font-family: inherit;
  font-size: 1rem;
  font-weight: 600;
  margin-left: 1em;
  padding: 0.5em 1em;
  border: none;
  border-radius: 6px;
  background-color: #3078ca;
  color: white;
  outline: none;
}

.Button:not(:disabled) {
  cursor: pointer;
}

.Button:hover:not(:disabled) {
  background-color: #22548e;
}

.Button:disabled {
  opacity: 0.65;
}

.Button:focus {
  box-shadow: 0 0 0 2px white, 0 0 0 4px #3078ca;
}

.Message {
  margin-top: 3.5rem;
  font-weight: 600;
}

.Message__error {
  color: red;
}

.Loading {
  margin-top: 3.5rem;
  font-weight: 600;
}

.UserInfo {
  margin-top: 3.5rem;
}

.Avatar {
  height: auto;
  vertical-align: middle;
}

.Avatar--user {
  border-radius: 50%;
}

.RepoList {
  margin-top: 3.5rem;
}

.RepoList__pagination {
  float: right;
}

.RepoList__pagination--top {
  line-height: 2rem;
  position: relative;
  top: 1px;
}

@media (max-width: 40em) {
  .RepoList__pagination {
    float: none;
  }

  .RepoList__pagination--top {
    margin-top: 0.83em;
  }
}

.Pagination {
  font-size: 1.17rem;
  font-weight: 400;
  white-space: nowrap;
}

.Pagination__link {
  font-weight: 600;
}

.Pagination__link--disabled {
  pointer-events: none;
  opacity: 0.65;
}

.Pagination__link.Link:hover {
  text-decoration: none;
  color: #024ca0;
}

.RepoInfo + .RepoInfo {
  margin-top: 2.5rem;
}

.HistoryRow__timestamp {
  opacity: 0.65;
}

.LargeHeading {
  font-size: 1.5rem;
  font-weight: 600;
  margin: 0.83em 0;
}

.SmallHeading {
  font-size: 1.17rem;
  font-weight: 700;
  margin: 0.83em 0;
}

.SmallHeading--largeMarginTop {
  margin-top: 1.67em;
}

.Row {
  display: flex;
  line-height: 1.5em;
  padding: 0.5em 0;
  position: relative;
}

.Row:nth-child(even)::before {
  content: " ";
  position: absolute;
  left: -1rem;
  right: -1rem;
  top: 0;
  bottom: 0;
  pointer-events: none;
  background-color: hsla(198, 45%, 10%, 0.04);
}

.Row__label {
  flex: 0 0 33%;
  min-width: 10.5em;
  margin-right: 1.5em;
}

.Row__content {
  flex: 1;
  margin-left: 0;
  font-weight: 600;
}

@media (max-width: 40em) {
  .Row {
    flex-direction: column;
  }

  .Row__label {
    margin-bottom: 0.25em;
  }
}

.Link {
  color: #0366d6;
  text-decoration: none;
  white-space: nowrap;
}

.Link:hover {
  text-decoration: underline;
}
