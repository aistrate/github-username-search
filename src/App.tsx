import {
  BrowserRouter as Router,
  NavLink,
  Switch,
  Route,
  Redirect,
  useLocation,
} from "react-router-dom";
import styled from "styled-components/macro";
import { ErrorBoundary } from "./Shared/ErrorBoundary";
import SearchPage from "./Search/SearchPage";
import HistoryPage from "./History/HistoryPage";
import AboutPage from "./About/AboutPage";

export default App;

function App() {
  return (
    <AppContainer>
      <AppHeader>{process.env.REACT_APP_APPLICATION_NAME}</AppHeader>

      <Router>
        <Nav />
        <Content />
      </Router>
    </AppContainer>
  );
}

function Nav() {
  return (
    <nav>
      <NavList>
        <NavItem to="/search">Search</NavItem>
        <NavItem to="/history">History</NavItem>
        <NavItem to="/about">About</NavItem>
      </NavList>
    </nav>
  );
}

function Content() {
  const location = useLocation();
  const queryParams = new URLSearchParams(location.search);

  return (
    <ContentContainer>
      <ErrorBoundary key={location.pathname}>
        <Switch>
          <Route path="/search">
            <SearchPage
              queryUsername={queryParams.get("q")}
              queryPage={queryParams.get("page")}
            />
          </Route>
          <Route path="/history">
            <HistoryPage />
          </Route>
          <Route path="/about">
            <AboutPage />
          </Route>
          <Route path="*">
            <Redirect to="/search" />
          </Route>
        </Switch>
      </ErrorBoundary>
    </ContentContainer>
  );
}

type NavItemProps = {
  to: string;
  children?: React.ReactNode;
};

function NavItem({ to, children }: NavItemProps) {
  return (
    <NavItemContainer>
      <StyledNavLink to={to} activeClassName="active">
        {children}
      </StyledNavLink>
    </NavItemContainer>
  );
}

const AppContainer = styled.div`
  max-width: 50rem;
  min-width: 20rem;
  min-height: 100vh;
  margin: auto;
`;

const AppHeader = styled.div`
  background-color: #1b4371;
  color: white;
  font-size: 2.25rem;
  text-align: center;
  padding: 0.6em 0;

  @media (max-width: 50em) {
    & {
      font-size: calc(1.25rem + 2vw);
    }
  }
`;

const NavList = styled.ul`
  list-style-type: none;
  margin: 0;
  padding: 0;
  overflow: hidden;
  background-color: rgb(0, 0, 0, 0.85);
  font-size: 1.25rem;
`;

const NavItemContainer = styled.li`
  float: left;
`;

const StyledNavLink = styled(NavLink)`
  display: block;
  color: rgb(255, 255, 255, 0.5);
  font-weight: 600;
  padding: 0.4em 0.8em;
  text-align: center;
  text-decoration: none;

  :hover {
    color: white;
    background-color: rgb(255, 255, 255, 0.1);
  }

  &.active {
    color: white;
  }
`;

const ContentContainer = styled.div`
  padding: 1rem;
  overflow-x: hidden;
`;
