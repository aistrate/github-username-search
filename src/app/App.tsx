import { Navigate, Route, Routes, useLocation } from "react-router-dom";
import styled from "styled-components/macro";
import ErrorBoundary from "../common/ErrorBoundary";
import AboutPage from "../features/about/AboutPage";
import HistoryPage from "../features/history/HistoryPage";
import SearchPage from "../features/search/SearchPage";
import Nav from "./Nav";

export default App;

function App() {
  return (
    <AppContainer>
      <Header>{process.env.REACT_APP_APPLICATION_NAME}</Header>
      <Nav />
      <Page />
    </AppContainer>
  );
}

function Page() {
  const location = useLocation();
  const queryParams = new URLSearchParams(location.search);

  return (
    <PageContainer data-testid="pageContainer">
      <ErrorBoundary key={location.pathname}>
        <Routes>
          <Route
            path="/search"
            element={
              <SearchPage
                queryUsername={queryParams.get("username")}
                queryPage={queryParams.get("page")}
              />
            }
          />
          <Route path="/history" element={<HistoryPage />} />
          <Route path="/about" element={<AboutPage />} />
          <Route path="*" element={<Navigate to="/search" replace />} />
        </Routes>
      </ErrorBoundary>
    </PageContainer>
  );
}

const AppContainer = styled.div`
  max-width: 50rem;
  min-width: 20rem;
  min-height: 100vh;
  margin: auto;
`;

const Header = styled.div`
  background-color: #1b4371;
  color: white;
  font-size: 2.25rem;
  text-align: center;
  padding: 0.6em 0;

  @media (max-width: 50em) {
    & {
      font-size: calc(1.25rem + 2vw);
    }
  }
`;

const PageContainer = styled.div`
  padding: 1rem;
  overflow-x: hidden;
`;
