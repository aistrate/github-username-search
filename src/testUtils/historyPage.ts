import { queryAllByRole, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

export function isHistoryPage() {
  return screen.queryByText(headingRegex) !== null;
}

export function headingCount() {
  const headingText = screen.getByText(headingRegex).textContent || "";
  const match = headingText.match(headingRegex) as RegExpMatchArray;

  return parseInt(match[1]);
}

export function historyUsernames() {
  return historyUsernameLinks().map((link) => link.textContent || "");
}

export function followUsernameLink(username: string) {
  const usernameLink = historyUsernameLinks().find(
    (link) => link.textContent === username
  );

  if (!usernameLink) {
    throw new Error(`Unable to find history link for username: ${username}`);
  }

  userEvent.click(usernameLink);
}

function historyUsernameLinks() {
  return queryAllByRole(screen.getByTestId("historyList"), "link");
}

const headingRegex = /History \((\d+)\)/;
