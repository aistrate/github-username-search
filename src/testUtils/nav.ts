import { getByText, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

export function goToSearchPage() {
  userEvent.click(getByText(menu(), "Search"));
}

export function goToHistoryPage() {
  userEvent.click(getByText(menu(), "History"));
}

export function goToAboutPage() {
  userEvent.click(getByText(menu(), "About"));
}

function menu() {
  return screen.getByRole("navigation");
}
