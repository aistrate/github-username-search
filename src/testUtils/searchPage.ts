import { getByText, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

// SearchForm

export function searchUsername(username: string) {
  typeUsername(username);
  userEvent.click(searchButton());
}

export function typeUsername(username: string) {
  userEvent.type(usernameInput(), username);
}

export function clearUsername() {
  userEvent.clear(usernameInput());
}

export function usernameInputValue() {
  return usernameInput().getAttribute("value");
}

export function usernameInputHasFocus() {
  return usernameInput() === document.activeElement;
}

function usernameInput() {
  return screen.getByPlaceholderText("Username");
}

function searchButton() {
  return screen.getByRole("button", { name: "Search" });
}

// UserInfo

export function userInfoText() {
  return userInfo().textContent || "";
}

export async function waitForUserInfo() {
  await screen.findByTestId("userInfo");
}

function userInfo() {
  return screen.getByTestId("userInfo");
}

// Pagination

export function goToNextRepoPage() {
  userEvent.click(getByText(topPagination(), /next/i));
}

export function goToPreviousRepoPage() {
  userEvent.click(getByText(topPagination(), /previous/i));
}

export function paginationText() {
  return topPagination().textContent || "";
}

export async function waitForPaginationControl() {
  await screen.findByTestId("topPagination");
}

function topPagination() {
  return screen.getByTestId("topPagination");
}
