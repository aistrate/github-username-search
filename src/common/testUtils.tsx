import { render } from "@testing-library/react";
import { ReactElement, ReactNode } from "react";
import { Provider } from "react-redux";
import { MemoryRouter, useLocation } from "react-router-dom";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";
import styled from "styled-components";
import { createStore } from "../app/store";

export {
  renderWithWrapper,
  createRendererWithWrapper,
  delay,
  RoutingLocation,
  removeClassNames,
};

/**
 * Renderer used by most tests. Uses jsdom as a full DOM implementation in Node.js.
 * Can handle user events (click, etc.) on the page.
 */
function renderWithWrapper(ui: ReactElement, initialRoute = "/") {
  const Wrapper = createWrapper(initialRoute);

  return render(ui, { wrapper: Wrapper });
}

/**
 * Renderer used by snapshot tests.
 */
function createRendererWithWrapper(ui: ReactElement, initialRoute: string) {
  const Wrapper = createWrapper(initialRoute);

  return renderer.create(<Wrapper>{ui}</Wrapper>);
}

function createWrapper(initialRoute: string) {
  const store = createStore();

  const Wrapper = ({ children }: { children: ReactNode }) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={[initialRoute]}>{children}</MemoryRouter>
    </Provider>
  );

  return Wrapper;
}

function delay(ms: number) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

function RoutingLocation() {
  const location = useLocation();

  return (
    <Container data-testid="routingLocation">
      {location.pathname + location.search}
    </Container>
  );
}

const Container = styled.div`
  position: fixed;
`;

/**
 * CSS class names generated by styled-components are non-deterministic,
 * so they will interfere with snapshot testing.
 */
function removeClassNames(
  tree: null | ReactTestRendererJSON | ReactTestRendererJSON[]
) {
  if (!tree) return;

  if (!Array.isArray(tree)) {
    tree = [tree];
  }

  tree.forEach((n) => {
    if (typeof n === "object" && n.type) {
      if (n.props && n.props.className) {
        delete n.props.className;
      }

      if (n.children) {
        n.children.forEach((child) => {
          if (typeof child !== "string") {
            removeClassNames(child);
          }
        });
      }
    }
  });
}
