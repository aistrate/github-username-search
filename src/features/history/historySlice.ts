import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { AppDispatch, RootState } from "../../app/store";
import { getLocalStorageItem } from "../../common/localStorage";

type HistoryItem = {
  username: string;
  timestamp: number;
};

const localStorageKey = "searchHistory";

const loadHistory = () => (dispatch: AppDispatch) => {
  const history = getLocalStorageItem<HistoryItem[]>(localStorageKey, []);

  return dispatch(setHistory(history));
};

const initialState = null as HistoryItem[] | null;

const historySlice = createSlice({
  name: "history",
  initialState,
  reducers: {
    setHistory(_state, action: PayloadAction<HistoryItem[]>) {
      return action.payload;
    },
    resetHistory() {
      return initialState;
    },
  },
});

const selectHistory = (state: RootState) => state.history;

export default historySlice.reducer;
export const { setHistory, resetHistory } = historySlice.actions;

export type { HistoryItem };
export { loadHistory, selectHistory };
