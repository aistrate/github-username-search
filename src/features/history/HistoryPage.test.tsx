import { screen } from "@testing-library/react";
import App from "../../app/App";
import * as historyPage from "../../testUtils/historyPage";
import * as nav from "../../testUtils/nav";
import * as searchPage from "../../testUtils/searchPage";
import { renderWithWrapper } from "../../testUtils/utils";
import { HistoryItem } from "./historySlice";

test("on every successful search, save username to the top of search history", async () => {
  renderWithWrapper(<App />);

  nav.goToHistoryPage();
  expectHistoryToEqual([]);

  nav.goToSearchPage();
  searchPage.searchUsername("reddit");
  await searchPage.waitForUserInfo();

  nav.goToHistoryPage();
  expectHistoryToEqual(["reddit"]);

  nav.goToSearchPage();
  searchPage.searchUsername("graphql");
  await searchPage.waitForUserInfo();

  nav.goToHistoryPage();
  expectHistoryToEqual(["graphql", "reddit"]);

  nav.goToSearchPage();
  searchPage.searchUsername("reddit");
  await searchPage.waitForUserInfo();

  nav.goToHistoryPage();
  expectHistoryToEqual(["reddit", "graphql"]);

  nav.goToSearchPage();
  searchPage.searchUsername("nonexistent");
  expect(
    await screen.findByText("Username 'nonexistent' was not found.")
  ).toBeInTheDocument();

  nav.goToHistoryPage();
  expectHistoryToEqual(["reddit", "graphql"]);
});

test("username on History page links to Search page for that username", async () => {
  renderWithWrapper(<App />);

  nav.goToSearchPage();
  searchPage.searchUsername("graphql");
  await searchPage.waitForUserInfo();

  nav.goToHistoryPage();
  expect(historyPage.headingCount()).toBe(1);

  historyPage.followUsernameLink("graphql");
  expect(searchPage.usernameInputValue()).toBe("graphql");

  const repoHeadings = await screen.findAllByRole("heading", { level: 3 });
  expect(repoHeadings.length).toBe(25);
});

test("convert username to lowercase before saving it to search history", async () => {
  renderWithWrapper(<App />);

  nav.goToSearchPage();
  searchPage.searchUsername("GraphQL"); // capitalized
  await searchPage.waitForUserInfo();

  nav.goToHistoryPage();
  expect(historyPage.historyUsernames()).not.toContain("GraphQL");
  expect(historyPage.historyUsernames()).toContain("graphql");
});

test("empty History page shows '(empty)'", async () => {
  renderWithWrapper(<App />, "/history");

  expect(historyPage.headingCount()).toBe(0);
  expect(screen.getByTestId("historyList")).toHaveTextContent("(empty)");
});

test("render the History page with history data (snapshot test)", async () => {
  setHistory([
    { username: "reddit", timestamp: 1643299499644 },
    { username: "graphql", timestamp: 1643299495994 },
  ]);

  const { container } = renderWithWrapper(<App />, "/history");

  expect(container.firstChild).toMatchSnapshot();
});

function expectHistoryToEqual(expected: string[]) {
  expect(historyPage.historyUsernames()).toEqual(expected);
  expect(historyPage.headingCount()).toBe(expected.length);
}

function setHistory(history: HistoryItem[]) {
  window.localStorage.setItem("searchHistory", JSON.stringify(history));
}
