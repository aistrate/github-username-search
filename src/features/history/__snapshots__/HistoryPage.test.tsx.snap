// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the History page (snapshot test) 1`] = `
<div>
  <div>
    GitHub Username Search
  </div>
  <nav>
    <ul>
      <li>
        <a
          href="/search"
          onClick={[Function]}
        >
          Search
        </a>
      </li>
      <li>
        <a
          aria-current="page"
          href="/history"
          onClick={[Function]}
        >
          History
        </a>
      </li>
      <li>
        <a
          href="/about"
          onClick={[Function]}
        >
          About
        </a>
      </li>
    </ul>
  </nav>
  <div>
    <h2>
      History 
      (0)
    </h2>
    <div
      data-testid="historyList"
    >
      (empty)
    </div>
  </div>
</div>
`;
